runtime/lib/ttnn/operations/deletion/dealloc.cpp

```diff
@@ -6,10 +6,11 @@
 
 namespace tt::runtime::ttnn::operations::deletion {
 void run(const ::tt::target::ttnn::DeallocOp *op, ProgramContext &context) {
-  bool force = true;
   ProgramTensorPool &tensorPool = context.getTensorPool();
   ::ttnn::Tensor &tensor = tensorPool.at(op->in()->global_id());
-  tensor.deallocate(force);
+  assert(tensor.is_allocated());
+  tensor.deallocate();
+  assert(!tensor.is_allocated());
```

Comment:
Actually sorry this seems a bit weird to me as well. If we're not force deallocating, I don't think this assert is valid

Reply:
Well it definitely works all tests pass. I even debugged it in TTNN to confirm. Not all tensor objects will get their own buffer storage, some share it. For example reshape op will return a new Tensor object but storage is shared with input Tensor object because data is the same but its interpretation is different. Force dealloc of this tensor object will force it ignore refcount to storage and make other ops downstream to work with deallocated Tensor which will lead to memory corruptions and wrong results. Thats why all the asserts because we were just ignoring that our input is actually invalid, TTNN asserted when returning to host that our data is crap.

Reply:
Ok I see so the target of the dealloc op is to deallocate the local tensor instead of globally deallocating the data. I was under the assumption that dealloc ops were inserted once the compiler knows the tensor has no more references, and we should force deallocate in runtime.

Reply:
Yeah so given this, I don't think the assert is correct. If you look at is_allocated:
```
bool Tensor::is_allocated() const {
    ZoneScoped;
    auto output = std::visit(
        [](auto&& storage) -> bool {
            return storage.is_allocated();
        },
        this->get_storage());
    return output;
}
```

It checks the status of the storage, which is shared by every tensor that references the same data. Without force deallocating or ensuring the ref count == 1, this data may still be alive. It's just a coincidence that all tests are passing - which means every tensor just has 1 reference:
```
                     if ((force or ref_count_to_use == 1) and not this->tensor_attributes->deallocated) {
                            this->tensor_attributes->deallocated = true;
                            auto dealloc_lambda = std::make_shared<std::function<void(Device*)>>(
                                [force, attr = this->tensor_attributes](Device* worker) mutable {
                                    ZoneScopedN("ShardDeallocate");
                                    TT_ASSERT(std::holds_alternative<tt::tt_metal::MultiDeviceStorage>(attr->storage), "Unexpected type {}", tt::stl::get_active_type_name_in_variant(attr->storage));
                                    auto& s = std::get<MultiDeviceStorage>(attr->storage);
                                    if (s.has_buffer_for_device(worker)) {
                                        auto& device_buffer = s.get_buffer_for_device(worker);
                                        if (force or device_buffer.use_count() == 1) {
                                           // ******** Storage deallocation happens only when force is set or ref count is 1 //
                                            DeallocateBuffer(*device_buffer);
                                        }
                                        device_buffer.reset();
                                    }
                                });
```
But then it's weird that your seeing errors when force deallocation is set to true.

Reply:
I confirmed that refcount is 2 in case above from debugger. This likely means that storage as an object is abstracted above real data as well. Maybe based on storage type.

Reply:
Correction:refcount was 1, buffer use_count was 2.